homeassistant/components/sensor/mqtt.py

```diff
@@ -128,6 +132,8 @@ def message_received(topic, payload, qos):
             if self._template is not None:
                 payload = self._template.async_render_with_possible_json_value(
                     payload, self._state)
+            else:
+                payload = payload[:254]
```

Comment:
Excuse my language, but why the heck should we limit the payload size to 255 characters?! Some people might depend on it, and it would just be a very weird quirk of this platform. We need to follow the MQTT spec or we'll have to deal with lots of issues down the line.

Reply:
The problem is that json payloads are often / almost always above the 255 character limit for HA states. Then setting a too long json payload will trigger an exception, without the attributes being set.

Reply:
mqtt max payload size should be around 268435455 bytes by the way.

Reply:
Oh I didn't know about the limitation of 255 characters for states. I'm unsure though whether we should just "cut off" the last bit then. Failing silently like this would do is prone to issues.

Consider this: somebody might be parsing the states of MQTT sensors at a later time. For example they could store the raw json in the state and only be parsing it at a later point with templates. Just chopping off the last characters would break a json parser.

I do see the issue about payloads frequently causing exceptions. But then we should definitely not just chop off the last bit as that might cause lots of issues. Maybe we can check if the payload is longer than 255 characters and then just not report the state? But that would again have the issue that we would be failing silently, also not very nice.